include/llbuild/Core/BuildEngine.h

```diff
@@ -34,8 +34,39 @@
 namespace llbuild {
 namespace core {
 
-// FIXME: Need to abstract KeyType;
-typedef std::string KeyType;
+struct KeyType {
+  std::string function;
+  std::vector<uint8_t> args;
+
+  bool operator==(const KeyType& rhs) const {
+    return function == rhs.function && args == rhs.args;
+  }
+
+  bool operator<(const KeyType& rhs) const {
+    return function <= rhs.function && args < rhs.args;
+  }
+
+  std::string describe() const {
+    // FIXME: encode args
+    return function;
```

Comment:
This seems important? :)

Reply:
What is `describe` for? If it is for debugging, LLVM has a convention of "dump" methods that write to a raw_ostream, I'd somewhat rather follow that than unnecessary heap allocations.

Reply:
From later, it looks like describe is not just for debugging, so definitely deserves a doc comment explaining it.

Reply:
(and also, still, I think I would prefer that it write to a raw_ostream so we avoid a lot of heap allocations in using it).

This will require various refactors in the clients, which does make me think it might be worth doing a refactor to switch KeyType to its own struct, add a bunch of methods, and then stage in the function/args split later.